UniqueArrayImp.h

```diff
@@ -0,0 +1,116 @@
+#ifndef MTMPARKINGLOT_UNIQUEARRAYIMP_H
+#define MTMPARKINGLOT_UNIQUEARRAYIMP_H
+
+template <class Element, class Compare>
+UniqueArray<Element, Compare>::UniqueArray(unsigned int size) :
+    data(new Element*[size]()), size(size) {}
+
+template <class Element, class Compare>
+UniqueArray<Element, Compare>::UniqueArray(const UniqueArray& other) :
+    data(new Element*[other.size]()), size(other.size)
+{
+    for (unsigned int i = 0; i < size; ++i) {
+        if (other.data[i]) {
+            data[i] = new Element(*other.data[i]);
+        }
+    }
+}
+
+template <class Element, class Compare>
+UniqueArray<Element, Compare>::~UniqueArray() {
+    for (unsigned int i = 0; i < size; ++i) {
+        if (data[i]) {
+            delete data[i];
+        }
+    }
+    delete[] data;
+}
+
+template <class Element, class Compare>
+unsigned int UniqueArray<Element, Compare>::insert(const Element& element) {
+    unsigned int index;
+    if (getIndex(element, index)) {
+        return index;
+    }
+    // else - new element:
+    for (unsigned int i = 0; i < size; ++i) {
+        if (!data[i]) {
+            data[i] = new Element(element);
+            return i;
+        }
+    }
+
+    throw UniqueArray::UniqueArrayIsFullException();
+}
+
+template <class Element, class Compare>
+bool UniqueArray<Element, Compare>::getIndex(const Element& element,
+    unsigned int& index) const
+{
+    Compare equal;
+    for (unsigned int i = 0; i < size; ++i) {
+        if (data[i] && equal(*data[i], element)) {
+            index = i;
+            return true;
+        }
+    }
+    return false;
+}
+
+template <class Element, class Compare>
+const Element* UniqueArray<Element, Compare>::operator[](
+    const Element& element) const
+{
+    unsigned int index;
+    if (getIndex(element, index)) {
+        return data[index];
+    }
+    return NULL;
+}
+
+template <class Element, class Compare>
+bool UniqueArray<Element, Compare>::remove(const Element& element) {
+    unsigned int index;
+    if (getIndex(element, index)) {
+        delete data[index];
```

Comment:
I think you're deleting a pointer here and not the actual element

Reply:
This confused me as well, but apparently, `delete` *must* receive a pointer to the element that it needs to delete.
This code should be definitely correct, it compiled and ran without faults and Valgrind gave it the all clear.